class.jetpack-post-images.php

```diff
@@ -339,21 +345,49 @@ static function from_html( $html_or_id ) {
 			return $images;
 		}
 
-		preg_match_all( '!<img.*src=[\'"]([^"]+)[\'"].*/?>!iUs', $html, $matches );
-		if ( !empty( $matches[1] ) ) {
-			foreach ( $matches[1] as $match ) {
-				if ( stristr( $match, '/smilies/' ) )
-					continue;
+		// Do not go any further if DOMDocument is disabled on the server.
+		if ( ! class_exists( 'DOMDocument' ) ) {
+			return $images;
+		}
 
-				$images[] = array(
-					'type'  => 'image',
-					'from'  => 'html',
-					'src'   => html_entity_decode( $match ),
-					'href'  => '', // No link to apply to these. Might potentially parse for that as well, but not for now
-				);
+		// Let's grab all image tags from the HTML.
+		$dom_doc = new DOMDocument;
+		$dom_doc->loadHTML( $html );
+		$image_tags = $dom_doc->getElementsByTagName( 'img' );
```

Comment:
It's great to see us ditch regexes and use an XML parser to go through HTML, but would it result in warnings for malformed HTML? Perhaps we need to switch to internal errors here as well?

Reply:
True. In 256ca47dc42a24559f3e3c0441302a4f62a62e30 and a6dff0855f365dc4b5bc4edddc82368d55d3ef8b I applied the changes we made to #7919